typed/merge.go

```diff
@@ -172,72 +172,83 @@ func (w *mergingWalker) visitListItems(t *schema.List, lhs, rhs value.List) (err
+		for i := 0; i < length; i++ {
+			child := list.At(i)
 			pe, err := listItemToPathElement(w.allocator, w.schema, t, i, child)
 			if err != nil {
-				errs = append(errs, errorf("lhs: element %v: %v", i, err.Error())...)
+				errs = append(errs, errorf("%s: element %v: %v", name, i, err.Error())...)
 				// If we can't construct the path element, we can't
 				// even report errors deeper in the schema, so bail on
 				// this element.
 				continue
 			}
-			if _, ok := observedLHS.Get(pe); ok {
-				errs = append(errs, errorf("lhs: duplicate entries for key %v", pe.String())...)
+			if _, _, found := observed.Get(pe); found {
+				errs = append(errs, errorf("%s: duplicate entries for key %v", name, pe.String())...)
+				continue
 			}
-			observedLHS.Insert(pe, child)
-			lhsOrder = append(lhsOrder, pe)
+			observed.Insert(pe, child, i)
+			pes = append(pes, pe)
 		}
+		return pes, observed, errs
 	}
 
-	// Then merge with RHS children.
-	observedRHS := fieldpath.MakePathElementSet(rLen)
-	if rhs != nil {
-		for i := 0; i < rhs.Length(); i++ {
-			child := rhs.At(i)
-			pe, err := listItemToPathElement(w.allocator, w.schema, t, i, child)
-			if err != nil {
-				errs = append(errs, errorf("rhs: element %v: %v", i, err.Error())...)
-				// If we can't construct the path element, we can't
-				// even report errors deeper in the schema, so bail on
-				// this element.
-				continue
-			}
-			if observedRHS.Has(pe) {
-				errs = append(errs, errorf("rhs: duplicate entries for key %v", pe.String())...)
-				continue
-			}
-			observedRHS.Insert(pe)
+	lhsOrder, observedLHS, lhsErrs := createPathElementsValues("lhs", lhs)
+	errs = append(errs, lhsErrs...)
+	rhsOrder, observedRHS, rhsErrs := createPathElementsValues("rhs", rhs)
+	errs = append(errs, rhsErrs...)
+
+	lLen, rLen = len(lhsOrder), len(rhsOrder)
+	for lI, rI := 0, 0; lI < lLen || rI < rLen; {
+		merge := func(pe fieldpath.PathElement, lChild, rChild value.Value) {
 			w2 := w.prepareDescent(pe, t.ElementType)
-			w2.rhs = child
-			if lchild, ok := observedLHS.Get(pe); ok {
-				w2.lhs = lchild
-			}
+			w2.lhs = lChild
+			w2.rhs = rChild
 			errs = append(errs, w2.merge(pe.String)...)
 			if w2.out != nil {
 				out = append(out, *w2.out)
 			}
 			w.finishDescent(w2)
 		}
-	}
-
-	for _, pe := range lhsOrder {
-		if observedRHS.Has(pe) {
+		if lI < lLen && rI < rLen && lhsOrder[lI].Equals(rhsOrder[rI]) {
+			// merge LHS & RHS items
+			pe := lhsOrder[lI]
+			lChild, _, _ := observedLHS.Get(pe)
+			rChild, _, _ := observedRHS.Get(pe)
+			merge(pe, lChild, rChild)
+			lI++
+			rI++
 			continue
 		}
-		value, _ := observedLHS.Get(pe)
-		w2 := w.prepareDescent(pe, t.ElementType)
-		w2.lhs = value
-		errs = append(errs, w2.merge(pe.String)...)
-		if w2.out != nil {
-			out = append(out, *w2.out)
+		if lI < lLen {
+			if _, index, ok := observedRHS.Get(lhsOrder[lI]); ok && index < rI {
+				// Skip the LHS item because it has already appeared
+				lI++
+				continue
```

Comment:
you can remove these continue (and above) if you properly use `else if`, no?

Reply:
That's the Return-early pattern. Should be preferred to the lots of `else if`s

Reply:
Yeah, I'm a big fan of early returns, but I don't think it really applies well here. It doesn't remove the extra-level of indentation.

Reply:
That's kind of a language-level limitation. I will need to test `iL < lLen` before I can get `ok` in the if statement.